servers/src/main/java/tachyon/master/block/BlockMaster.java

```diff
@@ -0,0 +1,648 @@
+        flushJournal();
+      }
+    }
+  }
+
+  /**
+   * @param blockId the block id to get information for
+   * @return the {@link BlockInfo} for the given block id. Called via RPC.
+   * @throws BlockInfoException
+   */
+  public BlockInfo getBlockInfo(long blockId) throws BlockInfoException {
+    synchronized (mBlocks) {
+      MasterBlockInfo masterBlockInfo = mBlocks.get(blockId);
+      if (masterBlockInfo == null) {
+        throw new BlockInfoException("Block info not found for " + blockId);
+      }
+      // Construct the block info object to return.
+      synchronized (mWorkers) {
+        return generateBlockInfo(masterBlockInfo);
+      }
+    }
+  }
+
+  /**
+   * Retrieves information for the given list of block ids. Called by internal masters.
+   *
+   * @param blockIds A list of block ids to retrieve the information for
+   * @return A list of {@link BlockInfo} objects corresponding to the input list of block ids. The
+   *         list is in the same order as the input list.
+   */
+  public List<BlockInfo> getBlockInfoList(List<Long> blockIds) {
+    List<BlockInfo> ret = new ArrayList<BlockInfo>(blockIds.size());
+    synchronized (mBlocks) {
+      synchronized (mWorkers) {
+        for (long blockId : blockIds) {
+          MasterBlockInfo masterBlockInfo = mBlocks.get(blockId);
+          if (masterBlockInfo != null) {
+            // Construct the block info object to return.
+            ret.add(generateBlockInfo(masterBlockInfo));
+          }
+        }
+        return ret;
+      }
+    }
+  }
+
+  /**
+   * @return the total bytes on each storage tier. Called by internal web ui.
+   */
+  public List<Long> getTotalBytesOnTiers() {
+    List<Long> ret = new ArrayList<Long>(Collections.nCopies(StorageLevelAlias.SIZE, 0L));
+    synchronized (mWorkers) {
+      for (MasterWorkerInfo worker : mWorkers) {
+        for (int i = 0; i < worker.getTotalBytesOnTiers().size(); i ++) {
+          ret.set(i, ret.get(i) + worker.getTotalBytesOnTiers().get(i));
+        }
+      }
+    }
+    return ret;
+  }
+
+  /**
+   * @return the used bytes on each storage tier. Called by internal web ui.
+   */
+  public List<Long> getUsedBytesOnTiers() {
+    List<Long> ret = new ArrayList<Long>(Collections.nCopies(StorageLevelAlias.SIZE, 0L));
+    synchronized (mWorkers) {
+      for (MasterWorkerInfo worker : mWorkers) {
+        for (int i = 0; i < worker.getUsedBytesOnTiers().size(); i ++) {
+          ret.set(i, ret.get(i) + worker.getUsedBytesOnTiers().get(i));
+        }
+      }
+    }
+    return ret;
+  }
+
+  /**
+   * Returns a worker id for the given worker. Returns the existing worker id if it exists. Called
+   * by workers, via RPC.
+   *
+   * @param workerNetAddress the worker {@link NetAddress}
+   * @return the worker id for this worker
+   */
+  public long getWorkerId(NetAddress workerNetAddress) {
+    // TODO: this NetAddress cloned in case thrift re-uses the object. Does thrift re-use it?
```

Comment:
I transferred the comment before, but it sounds like the fear is that thrift could re-use that object for the next request, thus mutating the object?